From a5ac0d27ea1906c58e35f03af8529b8829e3fead Mon Sep 17 00:00:00 2001
From: Steve Kondik <steve@cyngn.com>
Date: Mon, 30 Mar 2015 14:19:00 -0700
Subject: [PATCH 5/6] minui: Performance improvements

 * Add a blit operation that does no blending and use it.

Change-Id: I5c4ba91dd8683d91fe067a1b71c52da75e2d4569
---
 minui/Android.mk |  3 +--
 minui/graphics.c | 23 +++++++++++++++++++++++
 minui/minui.h    |  2 ++
 screen_ui.cpp    |  8 ++++----
 4 files changed, 30 insertions(+), 6 deletions(-)

diff --git a/minui/Android.mk b/minui/Android.mk
index ae60c52..d5d620c 100644
--- a/minui/Android.mk
+++ b/minui/Android.mk
@@ -45,11 +45,10 @@ include $(BUILD_STATIC_LIBRARY)
 
 include $(CLEAR_VARS)
 LOCAL_MODULE := libminui
-LOCAL_ARM_MODE:= arm
 LOCAL_SRC_FILES := $(common_src_files)
 LOCAL_ADDITIONAL_DEPENDENCIES := $(common_additional_dependencies)
 LOCAL_C_INCLUDES += $(common_c_includes)
-LOCAL_SHARED_LIBRARIES := libpng libpixelflinger
+LOCAL_SHARED_LIBRARIES := libpng
 LOCAL_CFLAGS += $(common_cflags) -DSHARED_MINUI
 LOCAL_WHOLE_STATIC_LIBRARIES := $(common_whole_static_libraries)
 include $(BUILD_SHARED_LIBRARY)
diff --git a/minui/graphics.c b/minui/graphics.c
index a4b5d48..88b0efa 100644
--- a/minui/graphics.c
+++ b/minui/graphics.c
@@ -314,6 +314,29 @@ void gr_blit(GRSurface* source, int sx, int sy, int w, int h, int dx, int dy) {
     unsigned char* src_p = source->data + sy*source->row_bytes + sx*source->pixel_bytes;
     unsigned char* dst_p = gr_draw->data + dy*gr_draw->row_bytes + dx*gr_draw->pixel_bytes;
 
+    for (int i = 0; i < h; ++i) {
+        memcpy(dst_p, src_p, w * source->pixel_bytes);
+        src_p += source->row_bytes;
+        dst_p += gr_draw->row_bytes;
+    }
+}
+
+void gr_blend(GRSurface* source, int sx, int sy, int w, int h, int dx, int dy) {
+    if (source == NULL) return;
+
+    if (gr_draw->pixel_bytes != source->pixel_bytes) {
+        printf("gr_blit: source has wrong format\n");
+        return;
+    }
+
+    dx += overscan_offset_x;
+    dy += overscan_offset_y;
+
+    if (outside(dx, dy) || outside(dx+w-1, dy+h-1)) return;
+
+    unsigned char* src_p = source->data + sy*source->row_bytes + sx*source->pixel_bytes;
+    unsigned char* dst_p = gr_draw->data + dy*gr_draw->row_bytes + dx*gr_draw->pixel_bytes;
+
     icon_blend_alpha(src_p, source->row_bytes, dst_p, gr_draw->row_bytes, w, h);
 }
 
diff --git a/minui/minui.h b/minui/minui.h
index 0e70036..7801a58 100644
--- a/minui/minui.h
+++ b/minui/minui.h
@@ -60,6 +60,8 @@ int gr_measure(const char *s);
 void gr_font_size(int *x, int *y);
 
 void gr_blit(gr_surface source, int sx, int sy, int w, int h, int dx, int dy);
+void gr_blend(gr_surface source, int sx, int sy, int w, int h, int dx, int dy);
+
 unsigned int gr_get_width(gr_surface surface);
 unsigned int gr_get_height(gr_surface surface);
 
diff --git a/screen_ui.cpp b/screen_ui.cpp
index f6345e2..f070f4b 100644
--- a/screen_ui.cpp
+++ b/screen_ui.cpp
@@ -168,18 +168,18 @@ void ScreenRecoveryUI::draw_progress_locked()
             if (rtl_locale) {
                 // Fill the progress bar from right to left.
                 if (pos > 0) {
-                    gr_blit(progressBarFill, width-pos, 0, pos, height, dx+width-pos, dy);
+                    gr_blend(progressBarFill, width-pos, 0, pos, height, dx+width-pos, dy);
                 }
                 if (pos < width-1) {
-                    gr_blit(progressBarEmpty, 0, 0, width-pos, height, dx, dy);
+                    gr_blend(progressBarEmpty, 0, 0, width-pos, height, dx, dy);
                 }
             } else {
                 // Fill the progress bar from left to right.
                 if (pos > 0) {
-                    gr_blit(progressBarFill, 0, 0, pos, height, dx, dy);
+                    gr_blend(progressBarFill, 0, 0, pos, height, dx, dy);
                 }
                 if (pos < width-1) {
-                    gr_blit(progressBarEmpty, pos, 0, width-pos, height, dx+pos, dy);
+                    gr_blend(progressBarEmpty, pos, 0, width-pos, height, dx+pos, dy);
                 }
             }
         }
-- 
1.9.1

